core/templates/services/playthrough.service.ts

```diff
@@ -1,334 +1,374 @@
+    const cycleWithNoCollision =
+    this.pathOfVisitedStates.slice(collisionIndex);
+    return [...cycleWithNoCollision, collision];
+  }
+
+  private currStateName(): string {
+    return this.pathOfVisitedStates[this.pathOfVisitedStates.length - 1];
+  }
+}
+
+class EarlyQuitTracker {
+  private stateName: string = ' ';
+  private expDurationInSecs: number | null = null;
+
+  foundAnIssue(): boolean {
+    return (
+      this.expDurationInSecs !== null &&
+      this.expDurationInSecs < ServicesConstants.EARLY_QUIT_THRESHOLD_IN_SECS);
+  }
+
+   recordExplorationQuit(
+       stateName: string,
+       expDurationInSecs: number | null
+   ): void {
+    if (this.stateName !== null && this.expDurationInSecs !== null) {
+      this.stateName = stateName;
+      this.expDurationInSecs = expDurationInSecs;
+    }
+  }
+
+
+  generateIssueCustomizationArgs(): EarlyQuitCustomizationArgs {
+    if (this.expDurationInSecs !== null) {
+      return {
+        state_name!: {value: this.stateName},
+        time_spent_in_exp_in_msecs!: {value: this.expDurationInSecs * 1000},
+      };
+    } else {
+      throw Error('Value expDurationInSecs should not be null!');
+    }
+  }
+}
+
+class MultipleIncorrectAnswersTracker {
+  private currStateName: string;
+  private numTries: number;
+
+  constructor(initStateName: string) {
+    this.currStateName = initStateName;
+    this.numTries = 0;
+  }
+
+  foundAnIssue(): boolean {
+    return this.numTries >= ServicesConstants.NUM_INCORRECT_ANSWERS_THRESHOLD;
+  }
+
+  recordStateTransition(destStateName: string): void {
+    if (this.currStateName === destStateName) {
+      this.numTries += 1;
+    } else {
+      this.currStateName = destStateName;
+      this.numTries = 0;
+    }
+  }
+
+  generateIssueCustomizationArgs(
+  ): MultipleIncorrectSubmissionsCustomizationArgs {
+    return {
+      state_name: {value: this.currStateName},
+      num_times_answered_incorrectly: {value: this.numTries},
+    };
+  }
+}
+
+ @Injectable({
+   providedIn: 'root'
+ })
+export class PlaythroughService {
+  private explorationId: string = ' ';
+  private explorationVersion: number | null = null;
+  private learnerIsInSamplePopulation: boolean | null = null;
+
+  private eqTracker: EarlyQuitTracker | null = null;
+  private cstTracker: CyclicStateTransitionsTracker | null = null;
+  private misTracker: MultipleIncorrectAnswersTracker | null = null;
+  private recordedLearnerActions: LearnerAction[] = [];
+  private playthroughStopwatch: Stopwatch | null = null;
+  private playthroughDurationInSecs: number | null = null;
+
+  constructor(
+       private explorationFeaturesService: ExplorationFeaturesService,
+       private learnerActionObjectFactory: LearnerActionObjectFactory,
+       private playthroughBackendApiService: PlaythroughBackendApiService,
+       private playthroughObjectFactory: PlaythroughObjectFactory) {}
+
+  initSession(
+      explorationId: string, 
+      explorationVersion: number | null,

```

Comment:
Why can this be `null`?

Reply:
I have assigned it null explicitely above in the code so as not to initialise it with a value.